### PR TITLE
Update CA1712: Add configuration for the number of values to trigger …

### LINKED ISSUE #2090: Add configuration option for CA1712 to control sensitivity
The current implementation follows the old FxCop rule's behavior, which will only fire if at least 75% of an enum's values appear to be prefixed with the name of the enum type. A config option could support three different thresholds:

1. Fire if _any_ field that starts with enum name
2. Fire only if _all_ the field names start with enum name
3. Default FxCop heuristic (75%)

Initially proposed in pull request 2089.

### PR DESCRIPTION
**Pull request by Evangelink**
…the rule

Add a configuration's option to enable different levels of strictness for the rule.

Fix #2090

### CODE REVIEW

File: docs/Analyzer Configuration.md [hunk 1]
```diff
@@ -38,7 +38,7 @@ This section documents the list of supported .editorconfig key-value options for
...
...m/visualstudio/code-quality/ca1721-property-names-should-not-match-get-methods), [CA1725](https://docs.microsoft.com/visualstudio/code-quality/ca1725-parameter-names-should-match-base-declaration), [CA1801](https://docs.microsoft.com/visualstudio/code-quality/ca1801-review-unused-parameters), [CA1802](https://docs.microsoft.com/visualstudio/code-quality/ca1802-use-literals-where-appropriate), [CA1815](https://docs.microsoft.com/visualstudio/code-quality/ca1815-override-equals-and-operator-equals-on-value-types), [CA1819](https://docs.microsoft.com/visualstudio/code-quality/ca1819-properties-should-not-return-arrays), [CA2217](https://docs.microsoft.com/visualstudio/code-quality/ca2217-do-not-mark-enums-with-flagsattribute), [CA2225](https://docs.microsoft.com/visualstudio/code-quality/ca2225-operator-overloads-have-named-alternates), [CA2226](https://docs.microsoft.com/visualstudio/code-quality/ca2226-operators-should-have-symmetrical-overloads), [CA2231](https://docs.microsoft.com/visualstudio/code-quality/ca2231-overload-operator-equals-on-overriding-valuetype-equals), [CA2234](https://docs.microsoft.com/visualstudio/code-quality/ca2234-pass-system-uri-objects-instead-of-strings)
```
Comment (Evangelink) [inline docs/Analyzer Configuration.md]: Adds CA1712 with the short url. Let me know if you prefer the other URL.

File: docs/Analyzer Configuration.md [hunk 2]
```diff
@@ -392,3 +392,20 @@ Option Values: integral values
 Default Value: Specific to each configurable rule ('100000' by default for most rules)
 
 Example: `dotnet_code_quality.CA5387.sufficient_IterationCount_for_weak_KDF_algorithm = 100000`
+
+#### Do not prefix enum values with type name
+Option Name: `enum_values_prefix_trigger`
```
Comment (Evangelink) [inline docs/Analyzer Configuration.md]: Feel free to suggest a better naming.

File: docs/Analyzer Configuration.md [hunk 3]
```diff
@@ -392,3 +392,20 @@ Option Values: integral values
...
+
+Configurable Rules: [CA1712](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1712)
+
+Option Values:
+
+| Option Value | Summary |
+| --- | --- |
+| `AnyEnumValue` | The rule will be triggered if _any_ of the enum values starts with the enum type name. |
```
Comment (Evangelink) [inline docs/Analyzer Configuration.md]: Also feel free to suggest better options and/or description.

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotPrefixEnumValuesWithTypeName.cs [hunk 4]
```diff
@@ -50,14 +51,19 @@ private static void AnalyzeNamedType(SymbolAnalysisContext context)
         {
             var symbol = (INamedTypeSymbol)context.Symbol;
             IEnumerable<ISymbol> enumValues;
-            if (symbol.TypeKind != TypeKind.Enum || !(enumValues = symbol.GetMembers().Where(m => m.Kind == SymbolKind.Field)).Any())
+            if (symbol.TypeKind != TypeKind.Enum || !(enumValues = symbol.GetMembers().Where(m => m.Kind == SymbolKind.Field && !m.IsImplicitlyDeclared)).Any())
```
Comment (Evangelink) [inline src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotPrefixEnumValuesWithTypeName.cs]: Excluding implicitly declared enums. This impacts the behavior of the current rule so if needed I could keep them for the 75% case.

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotPrefixEnumValuesWithTypeName.cs [hunk 5]
```diff
@@ -50,14 +51,19 @@ private static void AnalyzeNamedType(SymbolAnalysisContext context)
...
                 return;
             }
 
             var prefixedValues = enumValues.Where(m => m.Name.StartsWith(symbol.Name, StringComparison.OrdinalIgnoreCase));
             int percentPrefixed = 100 * prefixedValues.Count() / enumValues.Count();
-            if (percentPrefixed >= PercentValuesPrefixedThreshold)
+
+            var triggerOption = context.Options.GetEnumValuesPrefixTriggerOption(Rule, context.CancellationToken);
+
+            if (triggerOption == EnumValuesPrefixTrigger.AnyEnumValue ||
```
Comment (Evangelink) [inline src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotPrefixEnumValuesWithTypeName.cs]: Any preference between `if` and `switch`?

File: docs/Analyzer Configuration.md (hunk 3 above)
Comment (mavasani) [reply docs/Analyzer Configuration.md]: Looks good to me.

File: docs/Analyzer Configuration.md (hunk 1 above)
Comment (mavasani) [reply docs/Analyzer Configuration.md]: This list for rules that are configurable with `api_surface` option, which is unrelated to this PR. Please remove back CA1712 from this list.

File: docs/Analyzer Configuration.md [hunk 6]
```diff
@@ -392,3 +392,20 @@ Option Values: integral values
 Default Value: Specific to each configurable rule ('100000' by default for most rules)
 
 Example: `dotnet_code_quality.CA5387.sufficient_IterationCount_for_weak_KDF_algorithm = 100000`
+
+#### Do not prefix enum values with type name
```
Comment (mavasani) [inline docs/Analyzer Configuration.md]: You seem to have added the new option under configuration options for `Dataflow analysis`. You should move the added documentation above this section.

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotPrefixEnumValuesWithTypeNameTests.cs [hunk 7]
```diff
@@ -139,6 +140,144 @@ enum State
...
+                TestState =
+                {
+                    Sources =
+                    {
+                        @"
+                class A
+                {
+                    enum State
+                    {
+                        StateOk = 0,
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotPrefixEnumValuesWithTypeNameTests.cs]: You can probably add one more test where 3 out of the 4 fields have the prefix and validate the diagnostic fires for heuristic and any enum cases, but not the all enum values case.

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs [hunk 8]
```diff
@@ -40,6 +41,12 @@ internal static partial class AnalyzerOptionsExtensions
             CancellationToken cancellationToken)
             => options.GetNonFlagsEnumOptionValue(EditorConfigOptionNames.OutputKind, rule, s_defaultOutputKinds, cancellationToken);
 
+        public static EnumValuesPrefixTrigger GetEnumValuesPrefixTriggerOption(
+            this AnalyzerOptions options,
+            DiagnosticDescriptor rule,
+            CancellationToken cancellationToken)
```
Comment (mavasani) [inline src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: Might be good to match other API signatures in this file and have an additional parameter `EnumValuesPrefixTrigger defaultValue`, so each caller can decide it's default value.

File: docs/Analyzer Configuration.md (hunk 1 above)
Comment (Evangelink) [reply docs/Analyzer Configuration.md]: I didn't pay attention...Sorry

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs [hunk 9]
```diff
@@ -34,6 +35,13 @@ internal static partial class AnalyzerOptionsExtensions
             CancellationToken cancellationToken)
             => options.GetFlagsEnumOptionValue(EditorConfigOptionNames.RequiredModifiers, rule, defaultValue, cancellationToken);
 
+        public static EnumValuesPrefixTrigger GetEnumValuesPrefixTriggerOption(
+            this AnalyzerOptions options,
+            DiagnosticDescriptor rule,
+            EnumValuesPrefixTrigger defaultValue,
+            CancellationToken cancellationToken)
+            => options.GetFlagsEnumOptionValue(EditorConfigOptionNames.EnumValuesPrefixTrigger, rule, defaultValue, cancellationToken);
```
Comment (mavasani) [inline src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: Should this be using `GetNonFlagsEnumOptionValue`?

**Pull request merged**